### PR TITLE
Implement MapGenerator::generateIndustries

### PR DESCRIPTION
**Pull request by AaronVanGeffen**

### CODE REVIEW

File: src/OpenLoco/src/Map/MapGenerator/MapGenerator.cpp [hunk 1]
```diff
@@ -385,13 +387,83 @@ namespace OpenLoco::World::MapGenerator
...
+                        continue;
+                }
+                else if (numCargoAvailable == 0)
+                {
+                    continue;
+                }
+            }
+
+            const uint8_t numIndustriesFactor = S5::getOptions().numberOfIndustries;
+            const auto numIndustriesToCreate = IndustryManager::capOfTypeOfIndustry(indObjId, numIndustriesFactor);
```
Comment (AaronVanGeffen) [inline src/OpenLoco/src/Map/MapGenerator/MapGenerator.cpp:461]: This value is off. It seems to be half the size of what vanilla produces. Have not been able to narrow down the cause yet.

File: src/OpenLoco/src/Map/MapGenerator/MapGenerator.cpp [hunk 2]
```diff
@@ -416,6 +411,9 @@ namespace OpenLoco::World::MapGenerator
 
         CompanyManager::setUpdatingCompanyId(CompanyId::neutral);
 
+        auto progressTicksPerIndustry = (maxProgress - minProgress) / numIndustriesAvailable;
```
Comment (ZehMatt) [inline src/OpenLoco/src/Map/MapGenerator/MapGenerator.cpp:414]: Is there the possibility that `numIndustriesAvailable` can be zero here?

Comment (AaronVanGeffen) [reply src/OpenLoco/src/Map/MapGenerator/MapGenerator.cpp:414]: Nice catch. Yes, it ranges from 0-2. I should handle this differently.

Comment (AaronVanGeffen) [reply src/OpenLoco/src/Map/MapGenerator/MapGenerator.cpp:414]: Nice catch. Yes, though rarely, I should exit early in that case, I suppose.

File: src/OpenLoco/src/Map/MapGenerator/MapGenerator.cpp (hunk 1 above)
Comment (AaronVanGeffen) [reply src/OpenLoco/src/Map/MapGenerator/MapGenerator.cpp:534]: Turns out I have been chasing ghosts here. For a medium number of industries, master generates around 40 industries, and 60 industries for a high number. We're seeing similar numbers on this PR.

For some reason, I was expecting these numbers to be way closer to the kMaxIndustries limit, 128. I clearly thought I had seen this on master, but evidently, this is not the case.

**Comment by AaronVanGeffen** [comment]
Turns out I have been chasing ghosts here. For a medium number of industries, master generates around 40 industries, and 60 industries for a high number. We're seeing similar numbers on this PR.

For some reason, I was expecting these numbers to be way closer to the kMaxIndustries limit, 128. I clearly thought I had seen this on master, but evidently, this is not the case.

File: src/OpenLoco/src/Map/MapGenerator/MapGenerator.cpp [hunk 3]
```diff
@@ -433,13 +435,108 @@ namespace OpenLoco::World::MapGenerator
         }
     }
 
+    static void updateProgress(uint8_t value);
```
Comment (LeftofZen) [inline src/OpenLoco/src/Map/MapGenerator/MapGenerator.cpp:438]: can't you just move the method up here instead?

File: src/OpenLoco/src/Map/MapGenerator/MapGenerator.cpp [hunk 4]
```diff
@@ -433,13 +435,108 @@ namespace OpenLoco::World::MapGenerator
...
+        auto numIndustriesAvailable = 0;
+        for (auto indObjId = 0U; indObjId < ObjectManager::getMaxObjects(ObjectType::industry); indObjId++)
+        {
+            if (ObjectManager::get<IndustryObject>(indObjId) != nullptr)
+            {
+                numIndustriesAvailable++;
+            }
+        }
+
+        if (!numIndustriesAvailable)
```
Comment (LeftofZen) [inline src/OpenLoco/src/Map/MapGenerator/MapGenerator.cpp:482]: i wish there was a compiler option to disable implicit int-bool conversions like this. personally i'd write `if (numIndustriesAvailable == 0)`

File: src/OpenLoco/src/Map/MapGenerator/MapGenerator.cpp [hunk 5]
```diff
@@ -433,13 +435,108 @@ namespace OpenLoco::World::MapGenerator
...
+            if (industryObj == nullptr)
+            {
+                continue;
+            }
+
+            currentProgress += progressTicksPerIndustry;
+            updateProgress(currentProgress);
+
+            // Check if industry is available at present
+            if (getCurrentYear() < industryObj->designedYear || getCurrentYear() > industryObj->obsoleteYear)
```
Comment (LeftofZen) [inline src/OpenLoco/src/Map/MapGenerator/MapGenerator.cpp:502]: is this really `>` in the assembly? it makes more sense for it to be `>=`

File: src/OpenLoco/src/World/IndustryManager.cpp [hunk 6]
```diff
@@ -420,25 +420,27 @@ namespace OpenLoco::IndustryManager
     }
 
     // 0x00459722 & 0x004598F0
-    static int32_t capOfTypeOfIndustry(const uint8_t indObjId)
+    int32_t capOfTypeOfIndustry(const uint8_t indObjId, const uint8_t numIndustriesFactor)
```
Comment (LeftofZen) [inline src/OpenLoco/src/World/IndustryManager.cpp:423]: shouldn't this be static still

**Comment by LeftofZen** [comment]
seems to work nicely in playtesting, just a few small code comments

File: src/OpenLoco/src/Map/MapGenerator/MapGenerator.cpp (hunk 4 above)
Comment (AaronVanGeffen) [reply src/OpenLoco/src/Map/MapGenerator/MapGenerator.cpp:482]: Heh, I used an explicit comparison elsewhere as well. I'll change it here, too, for consistency.

File: src/OpenLoco/src/Map/MapGenerator/MapGenerator.cpp (hunk 3 above)
Comment (AaronVanGeffen) [reply src/OpenLoco/src/Map/MapGenerator/MapGenerator.cpp:438]: Heh, I just didn't want to clutter the diff with unrelated business. But it's a small method, so why not.

File: src/OpenLoco/src/Map/MapGenerator/MapGenerator.cpp (hunk 5 above)
Comment (AaronVanGeffen) [reply src/OpenLoco/src/Map/MapGenerator/MapGenerator.cpp:502]: You're right:
```asm
mov     cx, current_year
cmp     cx, [ebp+0CAh]  ; designedYear
jb      loc_4599A3      ; continue
cmp     cx, [ebp+0CCh]  ; obsoleteYear
jnb     loc_4599A3      ; continue
```
It's jnb (jump not below), so should be `>=`.

**Review by duncanspumpkin — APPROVED** [review]

**Pull request merged**